### PR TITLE
fix: handle runtime exceptions while establishing MQTT connection with IoT core

### PR DESCRIPTION
**Pull request by vaibhavmurkute**
**Issue #, if available:** 
-  RuntimeException thrown when it [fails to create MqttConnectionBuilder](https://github.com/aws-greengrass/aws-greengrass-nucleus/blob/main/src/main/java/com/aws/greengrass/mqttclient/MqttClient.java#L177) goes ignored.
- Possible `MqttException` thrown during [MqttClientConnection.connect()](https://awslabs.github.io/aws-crt-java/software/amazon/awssdk/crt/mqtt/MqttClientConnection.html#connect()), is currently not handled/logged : [link](https://github.com/aws-greengrass/aws-greengrass-nucleus/blob/main/src/main/java/com/aws/greengrass/mqttclient/AwsIotMqttClient.java#L251). 

**Description of changes:** 
- Wraps MqttConnectionProviderException with `MqttException` instead of generic `RuntimeException`. 
- Handles `MqttException`: Completes the future exceptionally if connection to IoT core fails with `MqttException`. 
- Exceptionally completed futures are logged automatically by nucleus.

**How was this change tested:** `mvn verify`

**Any additional information or context required to review the change:**

**Checklist:**
 - [ ] Updated the README if applicable
 - [ ] Updated or added new unit tests
 - [ ] Updated or added new integration tests
 - [ ] Updated or added new end-to-end tests
 - [ ] If your code makes a remote network call, it was tested with a proxy
 - [ ] You confirm that the change is backwards compatible

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/main/java/com/aws/greengrass/mqttclient/AwsIotMqttClient.java [hunk 1]
```diff
@@ -213,10 +214,14 @@ void reconnect() throws TimeoutException, ExecutionException, InterruptedExcepti
...
-                return disconnect();
-            });
+            try {
+                voidCompletableFuture = establishConnection(true).thenCompose((session) -> {
+                    initialConnect.set(false);
+                    return disconnect();
+                });
+            } catch (RuntimeException e) {
+                voidCompletableFuture.completeExceptionally(e);
+            }
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/mqttclient/AwsIotMqttClient.java:224]: instead, add the catch inside establishConnection and return a completed future which is completed exceptionally.

establishConnection is called twice, so your try catch won't catch everything.

Comment (vaibhavmurkute) [reply src/main/java/com/aws/greengrass/mqttclient/AwsIotMqttClient.java]: that makes sense, thanks. Fixed it.

**Review by wikimonkey — DISMISSED** [review]

**Review by wikimonkey — APPROVED** [review]

**Review by shaguptashaikh — APPROVED** [review]

File: src/test/java/com/aws/greengrass/mqttclient/AwsIotMqttClientTest.java [hunk 2]
```diff
@@ -128,6 +129,20 @@ void GIVEN_client_connection_fails_WHEN_subscribe_THEN_throws_exception(Extensio
...
+    @Test
+    void GIVEN_client_connection_throws_mqttexception_WHEN_subscribe_THEN_throws_exception() {
+        when(connection.connect()).thenThrow(MqttException.class);
+        doNothing().when(connection).onMessage(any());
+        when(builder.build()).thenReturn(connection);
+
+        AwsIotMqttClient client = new AwsIotMqttClient(() -> builder, (x) -> null, "A", mockTopic,
+                callbackEventManager, executorService, ses);
+        client.disableRateLimiting();
+        assertThrows(ExecutionException.class, () -> {
```
Comment (shaguptashaikh) [inline src/test/java/com/aws/greengrass/mqttclient/AwsIotMqttClientTest.java:141]: Nit - Could mock with a specific error message and assert on that message

**Comment by shaguptashaikh** [comment]
You will need to add this change to the release_2.5.x branch as well

**Pull request merged**